h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,178 @@
+package ai.h2o.targetencoding;
+
+import water.MRTask;
+import water.MemoryManager;
+import water.fvec.CategoricalWrappedVec;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+
+class BroadcastJoinForTargetEncoder {
+
+  static class FrameWithEncodingDataToArray extends MRTask<FrameWithEncodingDataToArray> {
+    
+    /**
+     * Numerators and denominators are being stored in the same row-array of the 2D array.
+     * 
+     *        _encodingDataPerNode[k][0 .. numOfCatLevels -1] will be storing numerators
+     *        _encodingDataPerNode[k][numOfCatLevels .. (2*numOfCatLevels -1)] will be storing denominators,
+     *        
+     *        where k is a current fold value (folds = {0,1,2,3...k}) or 0 when _foldColumnIdx == -1; 
+     */
+    int[][] _encodingDataPerNode = null;
+    
+    int _categoricalColumnIdx, _foldColumnIdx, _numeratorIdx, _denominatorIdx, _cardinalityOfCatCol, _maxFoldValue;
+
+    FrameWithEncodingDataToArray(int categoricalColumnIdx, int foldColumnId, int numeratorIdx, int denominatorIdx, int cardinalityOfCatCol, int maxFoldValue) {
+      _categoricalColumnIdx = categoricalColumnIdx;
+      _foldColumnIdx = foldColumnId;
+      _numeratorIdx = numeratorIdx;
+      _denominatorIdx = denominatorIdx;
+      _cardinalityOfCatCol = cardinalityOfCatCol;
+
+      if(foldColumnId == -1) {
+        _encodingDataPerNode = MemoryManager.malloc4(1, _cardinalityOfCatCol * 2);
+      } else {
+        assert maxFoldValue >= 1 : "There should be at leas two folds in the fold column";
+        assert _cardinalityOfCatCol > 0 && _cardinalityOfCatCol < (Integer.MAX_VALUE / 2)  : "Cardinality of categ. column should be within range (0, Integer.MAX_VALUE / 2 )";
+        _encodingDataPerNode = MemoryManager.malloc4(maxFoldValue + 1,_cardinalityOfCatCol * 2);
+      }
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+      Chunk categoricalChunk = cs[_categoricalColumnIdx];
+      Chunk numeratorChunk = cs[_numeratorIdx];
+      Chunk denominatorChunk = cs[_denominatorIdx];
+      for (int i = 0; i < categoricalChunk.len(); i++) {
+        int levelValue = (int) categoricalChunk.at8(i);
+
+        // We are allowed to do casting to `int` as we have validation before submitting this MRTask
+        int foldValue = _foldColumnIdx != -1 ? (int) cs[_foldColumnIdx].at8(i) : 0;
+        int[] arrForNumeratorsAndDenominators = _encodingDataPerNode[foldValue];
+
+        arrForNumeratorsAndDenominators[levelValue] = (int) numeratorChunk.at8(i);
+        arrForNumeratorsAndDenominators[_cardinalityOfCatCol + levelValue] = (int) denominatorChunk.at8(i);
+      }
+    }
+
+    @Override
+    public void reduce(FrameWithEncodingDataToArray mrt) {
+      int[][] leftArr = getEncodingDataArray();
+      int[][] rightArr = mrt.getEncodingDataArray();
+      
+      // Note: we need to add this check due to the following circumstances: 
+      //       1) there is a chance that reduce phase will start before map phase is finished. 
```

Comment:
Thanks, let's go into more details for future unknown reader:
For a given MR task T: we have T -> (Tl, Tr), and let's say Tl -> (Tll, Tlr).
Tll, Tlr `map` is always completed before Tl `reduce`, and T `reduce` needs to wait for completion of both Tl and Tr.
Tll and Tlr manipulate the same array, but not the same cells -> no race condition.
The race arises because Tl `reduce` (of both Tll and Tlr) can occur in parallel with Tr `map` (and `reduce`).
Once both Tl and Tr are completed, T `reduce` itself is safe.

Don't know how to summarize this, but it's far from being intuitive, so I think it's worth writing it explicitly.

Reply:
@seb-h2o 
> Don't know how to summarize this, but it's far from being intuitive, so I think it's worth writing it explicitly.

Agree. I just wonder whether these details should be here... it is actually quite general information and might be useful even outside of the BroadcastJoin scope. wdyt?

Reply:
@seb-h2o added more details to the comment to make it more explicit

Reply:
fantastic, thanks